ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+We relay on the ability to create namespaces directly. Which requires access to all authenticated users to have premissions to create a namespace.
+
+The boostrapping of the namespace created is delegate to a GitOps workflow that will automatically create an application with the necesary resources, including RoleBinding and Quotas and will continue to synch this through the lifecycle of the project.
+
+To be able to see exisiting workspaces for a specific user, we developed a small abstraction called workspace manager, which is able to determine what namepsace have a rolebinding providing access to the current user.
+
+There is currently no way to delete a workspace self-service. User will have to make a request to a konflux admin to delete a specific workspace
+
+### Namespace Limitations
+
+Kubernetes namespaces where not designed for self-service provisioning in a multi-tenant.
+
+Access for namespace list and create operation is all namepaces or none. This affects multi-tenancy if self-service is desired.
+
+Access to get and delete can be granted via RBAC once the namespace is created. This will required admin access which also affect self-service.
+
+Allowing access to specific Roles or ClusterRoles via RoleBinding can be done by an Admin, also braking self-service. This is required to be able to execute Konflux actions inside the namespace.
+
+## Decision
+
+Introduce a workspace CRD and corresponding controller as a project of our community.
+
+We will provide all authenticated users access to create their own workspaces within a limit.
+They will also be able to update, list, get and delete only workspaces they own. 
+
+## Consequences
+
+Extending k8s API with the workspace concept will allow konflux user to manage their konflux workspaces in a secured, self-service manner while konflux admin can maintain multi-tenancy.
+
+Such extention of the API that wraps namepsace has precedence, for example OpenShift apis provide a Project kind the wraps a k8s namespace.
+
+### Additional benefit 
+
+We can automate the bootstraping of namespaces with extra resources a konflux admin deems proper for their specific use case.
+
+We can create exension points that enable the community to use different backends of the implementation which will include the ability to scale to multiple cluster with a single instance of Konflux.
+
+
+## Konflux Workspaces Upstream Design Proposal
+
+### Purpose
+
+This a design proposal to represent the Konflux workspace concept long term. It aims to show how the Konflux upstream project should implement a workspace abstraction that will allow the community to select their own authentication provider as well as their own namespace provisioning backend. Three namespace provisioning backends will be presented as reference implementation but it should not be limited to those as the community will be able to extend or modify to better suit their use case.
```

Comment:
Can you please provide more details about the authentication provider,how is it related to workspaces?
Konflux upstream is already using [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/) and [dex](https://dexidp.io/) (keycloak can be used as well) for letting konflux admins configure social login. Upstream Konflux is not limited to using oauth2-proxy and dex, other implementation/services can be used as well.

Reply:
From my point of view we will support whatever authentication option k8s support: https://kubernetes.io/docs/reference/access-authn-authz/authentication/ whatever we have in place can work fine or we can think of a different authentication reference implementation as a separate topic. Having an abstraction for workspace does not affect authentication it does simplify authorization by auto generating the role binding for workspace creator.